### PR TITLE
Updated tests to include Exasol V8/ Update to vsjdbc 12.0.0 / Fix issues

### LINKED ISSUE #105: Update tests to include Exasol V8/ Update to vsjdbc 12.0.0 (VSEXA)
Since 2023-10-12 for version 8.23.0 of Exasol database a Docker image is available on [Dockerhub](https://hub.docker.com/layers/exasol/docker-db/8.23.0/images/sha256-c12588af953c73bcdd344d1b7cba6aea7fb709e09e97814664c4e16e87f77b73?context=explore)

The current ticket therefore requests to update the integration tests of VSEXA to use version 8.23.0 as latest default version.

Please note sibling-tickets for all [JDBC-based virtual schemas](https://github.com/exasol/virtual-schemas/blob/main/doc/user-guide/dialects.md).

### LINKED ISSUE #117: Fix javadoc generation
Currently javadoc is configured to not throw errors when failing in the current PR. 

The javadoc plugin in this project had custom configuration (found in pom.xml file): 
Javadoc fought with the ` udf-debugging-java` dependency when I added it, this dependency is needed to add support for remote debugging the VS integration tests. That's why I had to disable javadoc for now.

### PR DESCRIPTION
**Pull request by pj-spoelders**
Closes #105 
Closes #117

### CODE REVIEW

File: .github/workflows/dependencies_update.yml
Comment (kaklakariada) [inline .github/workflows/dependencies_update.yml:1]: Mark this as generated

File: .project-keeper.yml [hunk 1]
```diff
@@ -6,7 +6,11 @@ sources:
       - integration_tests
       - udf_coverage
       - maven_central
+excludes:
+  - "W-PK-CORE-153: Project-keeper version 4.1.0 is outdated. Please update project-keeper to latest version 4.2.0."
```
Comment (kaklakariada) [inline .project-keeper.yml:10]: There are no customizations here. Is it possible to upgrade to 4.2.0 or is there another problem?

File: doc/changes/changes_8.0.0.md [hunk 2]
```diff
@@ -0,0 +1,40 @@
+# Exasol Virtual Schema 8.0.0, released 2024-04-02
+
+Code name: Charset is always `utf-8`, deprecated IMPORT_DATA_TYPES `FROM_RESULT_SET` value .
```
Comment (kaklakariada) [inline doc/changes/changes_8.0.0.md:3]:
```suggestion
Code name: Charset is always `utf-8`, deprecated IMPORT_DATA_TYPES `FROM_RESULT_SET` value
```

File: doc/dialects/exasol.md [hunk 3]
```diff
@@ -113,6 +113,9 @@ Pushdown query with `GENERATE_JDBC_DATATYPE_MAPPING_FOR_EXA = 'true'`:
 ```sql
 IMPORT INTO (c1 DECIMAL(36,1), c2 .... ) FROM EXA AT "EXA_CONNECTION" STATEMENT '...'
 ```
+##### Data type mismatch
```
Comment (kaklakariada) [inline doc/dialects/exasol.md:116]:
```suggestion

##### Data type mismatch
```

add newline before heading

File: doc/dialects/exasol.md [hunk 4]
```diff
@@ -113,6 +113,9 @@ Pushdown query with `GENERATE_JDBC_DATATYPE_MAPPING_FOR_EXA = 'true'`:
...
+
+In case you run into a `Data type mismatch` issue when using EXA import turning this property on will usually also solve the issue by providing type hints.
```
Comment (kaklakariada) [inline doc/dialects/exasol.md:118]: Please add the exact error message and explicitly tell the user to set `GENERATE_JDBC_DATATYPE_MAPPING_FOR_EXA = 'true'`

File: pom.xml [hunk 5]
```diff
@@ -2,12 +2,13 @@
...
     <artifactId>exasol-virtual-schema</artifactId>
-    <version>7.2.0</version>
+    <version>8.0.0</version>
     <name>Exasol Virtual Schema</name>
     <description>This projects contains the Exasol dialect for Exasol's Virtual Schema</description>
     <url>https://github.com/exasol/exasol-virtual-schema/</url>
     <properties>
-        <vscjdbc.version>11.0.2</vscjdbc.version>
+        <vscjdbc.version>12.0.0</vscjdbc.version>
+        <org.testcontainers.version>1.19.0</org.testcontainers.version>
```
Comment (kaklakariada) [inline pom.xml:11]:
```suggestion
```

Seems to be not used

File: pom.xml [hunk 6]
```diff
@@ -160,13 +167,14 @@
             <plugin>
                 <groupId>org.apache.maven.plugins</groupId>
                 <artifactId>maven-javadoc-plugin</artifactId>
-                <version>3.4.1</version>
+                <version>3.6.3</version>
```
Comment (kaklakariada) [inline pom.xml:170]:
```suggestion
```

Version already specified in parent pom

File: src/main/java/com/exasol/adapter/dialects/exasol/ExasolLocalSqlGenerationVisitor.java [hunk 7]
```diff
@@ -0,0 +1,42 @@
+package com.exasol.adapter.dialects.exasol;
+
+import com.exasol.adapter.AdapterException;
+import com.exasol.adapter.dialects.SqlDialect;
+import com.exasol.adapter.dialects.rewriting.SqlGenerationContext;
+import com.exasol.adapter.metadata.DataType;
+import com.exasol.adapter.sql.SqlColumn;
+/**
+ * This class generates SQL queries for the {@link ExasolSqlDialect}. It is a 'specialisation' for LOCAL mode, based on {@link ExasolSqlGenerationVisitor}.
```
Comment (kaklakariada) [inline src/main/java/com/exasol/adapter/dialects/exasol/ExasolLocalSqlGenerationVisitor.java:9]: Please add a note how it is specialized.

File: src/main/java/com/exasol/adapter/dialects/exasol/ExasolLocalSqlGenerationVisitor.java
Comment (kaklakariada) [inline src/main/java/com/exasol/adapter/dialects/exasol/ExasolLocalSqlGenerationVisitor.java:1]: Check formatting of the this file, whitespace is missing.
Also add unit tests for this class.

File: src/main/java/com/exasol/adapter/dialects/exasol/ExasolLocalSqlGenerationVisitor.java [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+    public String visit(final SqlColumn column) throws AdapterException {
+        ExasolSqlDialect exasolSqlDialect = new ExasolSqlDialect(null,null);
+        String tablePrefix = "";
+        if (column.hasTableAlias()) {
+            tablePrefix = exasolSqlDialect.applyQuote(column.getTableAlias())
+                    + exasolSqlDialect.getTableCatalogAndSchemaSeparator();
+        } else if ((column.getTableName() != null) && !column.getTableName().isEmpty()) {
+            tablePrefix = exasolSqlDialect.applyQuote(column.getTableName())
+                    + exasolSqlDialect.getTableCatalogAndSchemaSeparator();
+        }
```
Comment (kaklakariada) [inline src/main/java/com/exasol/adapter/dialects/exasol/ExasolLocalSqlGenerationVisitor.java:31]: Extract this into a method

File: src/main/java/com/exasol/adapter/dialects/exasol/ExasolLocalSqlGenerationVisitor.java [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+        //PROVIDE A CAST TO CORRESPONDING DATATYPE WITH THE SAME LENGTH BUT WITH UTF8 AS CHARSET
```
Comment (kaklakariada) [inline src/main/java/com/exasol/adapter/dialects/exasol/ExasolLocalSqlGenerationVisitor.java:32]: Please DON'T SHOUT :)

File: src/main/java/com/exasol/adapter/dialects/exasol/ExasolLocalSqlGenerationVisitor.java [hunk 10]
```diff
@@ -0,0 +1,42 @@
...
+        if (column.getMetadata().getType().getExaDataType() == DataType.ExaDataType.CHAR || //
+                column.getMetadata().getType().getExaDataType() == DataType.ExaDataType.VARCHAR) {
+            int size = column.getMetadata().getType().getSize();
+            String dataTypeStr = column.getMetadata().getType().getExaDataType() == DataType.ExaDataType.CHAR ? "CHAR":"VARCHAR";
+            return "CAST(" + tablePrefix + exasolSqlDialect.applyQuote(column.getName())+" AS "+ dataTypeStr+"("+size+") UTF8)";
```
Comment (kaklakariada) [inline src/main/java/com/exasol/adapter/dialects/exasol/ExasolLocalSqlGenerationVisitor.java:37]: Extract `column.getMetadata().getType().getExaDataType()` into a local variable

File: src/test/java/com/exasol/adapter/dialects/exasol/AbstractExasolSqlDialectIT.java [hunk 11]
```diff
@@ -62,10 +63,26 @@ abstract class AbstractExasolSqlDialectIT {
     @BeforeAll
     static void beforeAll() throws BucketAccessException, TimeoutException, NoDriverFoundException, SQLException,
             FileNotFoundException {
```
Comment (kaklakariada) [inline src/test/java/com/exasol/adapter/dialects/exasol/AbstractExasolSqlDialectIT.java:65]:
```suggestion
    static void beforeAll() {
```

File: src/test/java/com/exasol/adapter/dialects/exasol/ExasolSqlDialectExaConnectionIT.java [hunk 12]
```diff
@@ -69,8 +70,19 @@ protected Map<String, String> getConnectionSpecificVirtualSchemaProperties() {
...
 
+    // These tests were overridden for the following reasons:
+    // Strict datatype checking which is by default enabled for Exasol V8 makes a number of these tests fail.
+    //
+    // The old tests that work for DB versions prior to strict datatype checking
+    // thus can only be run for earlier Exasol DB versions and have to be skipped for V8.
+    //
+    // The tests using the newer GENERATE_JDBC_DATATYPE_MAPPING_FOR_EXA switch that fixes most of the issues with strict
+    // datatype checking on the pushed down queries
+    // can be found in ExasolSqlDialectExaConnectionWithDataTypesIT .
```
Comment (kaklakariada) [inline src/test/java/com/exasol/adapter/dialects/exasol/ExasolSqlDialectExaConnectionIT.java:81]:
```suggestion
    // can be found in ExasolSqlDialectExaConnectionWithDataTypesIT.
```

**Review by kaklakariada — CHANGES_REQUESTED** [review]

File: .project-keeper.yml (hunk 1 above)
Comment (pj-spoelders) [reply .project-keeper.yml:10]: I'll contact you on slack about this

Comment (pj-spoelders) [reply .project-keeper.yml]: Updated PK

File: src/main/java/com/exasol/adapter/dialects/exasol/ExasolLocalSqlGenerationVisitor.java (hunk 8 above)
Comment (pj-spoelders) [reply src/main/java/com/exasol/adapter/dialects/exasol/ExasolLocalSqlGenerationVisitor.java:31]: Can you clarify this a bit further? Contacting you on slack.

File: src/main/java/com/exasol/adapter/dialects/exasol/ExasolLocalSqlGenerationVisitor.java (hunk 9 above)
Comment (pj-spoelders) [reply src/main/java/com/exasol/adapter/dialects/exasol/ExasolLocalSqlGenerationVisitor.java]: Sentence cased it ..

File: src/main/java/com/exasol/adapter/dialects/exasol/ExasolLocalSqlGenerationVisitor.java (hunk 10 above)
Comment (pj-spoelders) [reply src/main/java/com/exasol/adapter/dialects/exasol/ExasolLocalSqlGenerationVisitor.java]: Done.

File: src/test/java/com/exasol/adapter/dialects/exasol/AbstractExasolSqlDialectIT.java (hunk 11 above)
Comment (pj-spoelders) [reply src/test/java/com/exasol/adapter/dialects/exasol/AbstractExasolSqlDialectIT.java:65]: ok

File: doc/dialects/exasol.md (hunk 4 above)
Comment (pj-spoelders) [reply doc/dialects/exasol.md]: Done.

File: src/main/java/com/exasol/adapter/dialects/exasol/ExasolLocalSqlGenerationVisitor.java (hunk 7 above)
Comment (pj-spoelders) [reply src/main/java/com/exasol/adapter/dialects/exasol/ExasolLocalSqlGenerationVisitor.java]: Did that.

File: src/main/java/com/exasol/adapter/dialects/exasol/ExasolLocalSqlGenerationVisitor.java
Comment (pj-spoelders) [reply src/main/java/com/exasol/adapter/dialects/exasol/ExasolLocalSqlGenerationVisitor.java:1]: Formatting should be ok now. Working on unit tests.

File: src/main/java/com/exasol/adapter/dialects/exasol/ExasolLocalSqlGenerationVisitor.java [hunk 13]
```diff
@@ -4,39 +4,56 @@
...
+    ExasolLocalSqlGenerationVisitor(final SqlDialect dialect, final SqlGenerationContext context) {
         super(dialect, context);
     }
+
     @Override
     public String visit(final SqlColumn column) throws AdapterException {
-        ExasolSqlDialect exasolSqlDialect = new ExasolSqlDialect(null,null);
+        final ExasolSqlDialect exasolSqlDialect = new ExasolSqlDialect(null, null);
         String tablePrefix = "";
+        tablePrefix = getTablePrefix(column, exasolSqlDialect, tablePrefix);
```
[the remark this answers is not in the document]
Comment (pj-spoelders) [reply src/main/java/com/exasol/adapter/dialects/exasol/ExasolLocalSqlGenerationVisitor.java]: Cleaned it up a bit .. sloppy.

File: .github/workflows/dependencies_update.yml
Comment (pj-spoelders) [reply .github/workflows/dependencies_update.yml:1]: Done

**Review by kaklakariada — DISMISSED** [review]

File: pom.xml (hunk 5 above)
Comment (pj-spoelders) [reply pom.xml]: You're right. Removed it.

**Review by kaklakariada — APPROVED** [review]

**Pull request merged**